keps/sig-node/4210-max-image-gc-age/README.md

```diff
@@ -0,0 +1,373 @@
+
+### Non-Goals
+
+- Introduce other conditions for image garbage collection to trigger.
+    - A WG was put together in SIG-Node to collect other GC use cases, and no other conditions were identified.
+
+## Proposal
+
+Kubelet has three different configuration fields for image garbage collection:
+- ImageMinimumGCAge: the youngest an image can be to be qualified for garbage collection
+- ImageGCLowThresholdPercent: The lowest disk usage will be before garbage collection begins
+- ImageGCHighThresholdPercent: The highest disk usage will be before garbage collection runs each GC period.
+
+Between each of these options, there is a common thread: image garbage collection only triggers when disk usage has reached a certain threshold.
+In other words, there are no alternative conditions the Kubelet will begin triggering collection. Luckily, this condition is the most important:
+the primary goal of garbage collection is to ensure images don't clutter the disk too much and cause it to fill up needlessly.
+However, having the Kubelet be purely reactive means that images will clutter the disk and cause it to fill up. While there aren't reported cases
+where this causes issues, it is inefficient with disk space and can cause the Kubelet to scramble to save disk space when the threshold is met.
+
+An additional approach is to define a way for an admin to request images are cleaned up after they're unused for a certain period of time.
+This would reduce the frequency of the disk usage hitting the level, and provide an admin more flexibility in how garbage collection is defined.
+
+The proposal of this KEP is to add an option to the KubeletConfiguration object that looks like:
+```
+	// ImageMaximumGCAge is the maximum age an image can be unused before it is garbage collected.
+    // The default of this field is 0, which disables it.
+    // +optional
+	ImageMaximumGCAge metav1.Duration
+```
+
+To begin, this option will be set to 0, which will be interpreted as "disabled". In the future, a more reasonable default may be chosen.
+
+### User Stories (Optional)
+
+#### Story 1
+
+- As a cluster admin, I would like my unused images to be garbage collected in a timely manner, and not occupy disk space forever.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+- If set incorrectly the ImageMaximumGCAge option could cause unneeded image pulls. For instance, if a Cron job ran
+once a week, but the ImageMaximumGCAge was set to less than a week, that image would get pulled every week, causing needless
+traffic from the registry
+    - Proper documentation on this is the best way to mitigate this risk.
+    - Defining a good default for this value will be similarly tricky.
+
```

Comment:
Some additional risks I am thinking about based on comments above:

- Reliability of age detection. Can we start making mistake and try to delete something that is _mistakenly_ considered old (like `0` timestamp somewhere by mistake)
- If `ImageMaximumGCAge` is set to `0`, will anything break? I think some people may want it. So we need to check that it will not cause any reliability issues
  - For example. can there be any race between image was pulled, container scheduled, and GC removing the image?
- What is runtime behave incorrectly - can bad runtime implementation (e.g. not actually deleting the image) break (DDoS) the kubelet?
- Can anybody take a dependency on this behavior building a security-related feature? Can we guarantee anything about images being deleted or this is a best effort and this potential security feature will make bad assumptions?

Reply:
> Reliability of age detection. Can we start making mistake and try to delete something that is mistakenly considered old (like 0 timestamp somewhere by mistake)

since we're using the same mechanism as disk based currently does, I would imagine we'd have found these bugs already. it's possible though.

> For example. can there be any race between image was pulled, container scheduled, and GC removing the image?

that's a good point. Image GC period is currently 5 minutes, so maybe we need to make the minimum 5 minutes (both as a way of mitigating this and as a more realistic maximum: users may find the images are garbage collected slower than we seem to promise)

> What is runtime behave incorrectly - can bad runtime implementation (e.g. not actually deleting the image) break (DDoS) the kubelet?

I don't personally think the kubelet should be in the business of distrusting the runtime. The GC heartbeat is every 5 minutes so I think kubelet wouldn't get DoS'd , though it could be wasteful.

> Can anybody take a dependency on this behavior building a security-related feature? Can we guarantee anything about images being deleted or this is a best effort and this potential security feature will make bad assumptions?

I think we should mark it as best effort, because of the GC period and because kubelet doesn't control CRI behavior

Reply:
we should clearly state the best effort nature of this setting than - it will simplify the design.

As for not allowing anything less than 5 minutes - I am not sure about it. We can state that low values are also not a goal, but if this the case we need to pick something larger. So we are not on the edge of getting into this race condition. Maybe 30 minutes.

Not allowing low values has a side effect of not allowing this interesting scenario of never keeping image in the cache.